From e1329e6070893668bd97febabfe452a4f7169098 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Tue, 29 Jun 2021 06:58:55 +0300
Subject: [PATCH] [SW-46517] greco: provide full address of img_addr

In secureboot flow, the SDRAM addresses are defined from the ARC
viewpoint due to which img_addr is set as ARC viewpoint which is
not understood by LKD. This patch add changes to pass full address
of SRAM as part of img_addr.

Change-Id: Ie62a12151aa68ee6e36945996f70fb90d9782005
---

diff --git a/arch/greco_arc_gnu/common.c b/arch/greco_arc_gnu/common.c
index 7c88fd2..cbd7d12 100644
--- a/arch/greco_arc_gnu/common.c
+++ b/arch/greco_arc_gnu/common.c
@@ -76,6 +76,12 @@
 	return reg & mask;
 }
 
+u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	/* Return full address */
+	return (base | SDRAM_FULL_BASE);
+}
+
 u32 get_efuse_read_reg(void)
 {
 	return mmPSOC_EFUSE_READ;
diff --git a/preboot/module_lkd.c b/preboot/module_lkd.c
index b84ede2..7946254 100644
--- a/preboot/module_lkd.c
+++ b/preboot/module_lkd.c
@@ -439,6 +439,11 @@
 
 __weak void lkd_dev_spec_regs(struct cpu_dyn_regs *regs);
 
+__weak u64 lkd_dev_spec_get_img_addr(uintptr_t base)
+{
+	return base;
+}
+
 static int module_lkd_prep_desc(int size, u32 *offset)
 {
 	struct mem_desc mdesc, img_desc;
@@ -471,7 +476,7 @@
 	if (err)
 		return err;
 
-	cdesc->img_addr = img_desc.base;
+	cdesc->img_addr = lkd_dev_spec_get_img_addr(img_desc.base);
 
 	/* Header */
 	cdesc->header.magic = HL_COMMS_DESC_MAGIC;
